dds_web/api/schemas/project_schemas.py

```diff
@@ -84,3 +86,120 @@ def return_items(self, data, **kwargs):
+            models.File.project_id == sqlalchemy.func.binary(project.id)
+        )
+
+        # Get all files
+        files = all_contents_query.filter(models.File.name.in_(contents)).all()
+
+        # Get not found paths - may be folders
+        new_paths = set(contents).difference(x.name for x in files)
+
+        # Get all folder contents
+        folder_contents = {
+            x: all_contents_query.filter(models.File.subpath.like(f"{x.rstrip(os.sep)}%")).all()
+            for x in new_paths
+        }
+
+        # Not found
+        not_found = {x: folder_contents.pop(x) for x, y in list(folder_contents.items()) if not y}
+
+        return files, folder_contents, not_found
+
+    @marshmallow.post_dump
+    def return_items(self, data, **kwargs):
+        """Return project contents as serialized."""
+
+        # Fields
+        requested_items = data.get("requested_items")
+        url = data.get("url")
+        get_all = data.get("get_all")
+
+        # Check if project has contents
+        project_row = verify_project_exists(spec_proj=data.get("project"))
+        if not project_row.files:
+            raise ddserr.EmptyProjectException
+
+        # Check if specific files have been requested or if requested all contents
+        files, folder_contents, not_found = (None, None, None)
+        if requested_items:
+            files, folder_contents, not_found = self.find_contents(
+                project=project_row, contents=requested_items
+            )
+        elif get_all:
+            files = project_row.files
+        else:
+            raise ddserr.DDSArgumentError(message="No items were requested.")
+
+        # Items to return
+        found_files = {}
+        found_folder_contents = {}
+        not_found = {}
+
+        # Use file schema to get file info automatically
+        fileschema = sqlalchemyautoschemas.FileSchema(
+            many=False,
+            only=(
+                "name_in_bucket",
+                "subpath",
+                "size_original",
+                "size_stored",
+                "salt",
+                "public_key",
+                "checksum",
+                "compressed",
+            ),
+        )
+
+        # Connect to s3
+        with api_s3_connector.ApiS3Connector(project=project_row) as s3:
+            # TODO: Optimisation: Add check for if only searching for one file (head_bucket)
+            # Get bucket items
+            pages = s3.bucket_items()
+
+            # Get the info and signed urls for all files found in the bucket
+            for page in pages:
+                found_files.update(
+                    {
+                        x.name: {
+                            **fileschema.dump(x),
+                            "url": s3.generate_get_url(key=x.name_in_bucket) if url else None,
+                        }
+                        for x in files
+                        if x.name_in_bucket in page
+                    }
```

Comment:
If I understand this correctly, I think we can remove the check if a file is in the bucket or not. If a file that is in the database is not found in the bucket, that's a fairly serious error and we should raise an exception instead of silently filtering them out of the download list. I think you said the http request in the presigned URL would return 403 for non-existing files so I guess it would be easiest to raise that exception in the CLI for the user.

Reply:
Ok! I was thinking of doing something like that originally, I have no objections, will change.

Reply:
Done